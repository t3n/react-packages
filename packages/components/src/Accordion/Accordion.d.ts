import React, { ReactNode } from 'react';
import { MarginProps } from 'styled-system';
export interface AccordionProps extends MarginProps {
    title: string | JSX.Element;
    initialOpen?: boolean;
    children?: ReactNode;
}
declare const Accordion: React.FC<AccordionProps>;
export default Accordion;
