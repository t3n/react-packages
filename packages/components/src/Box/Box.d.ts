import { ColorProps, FlexboxProps, LayoutProps, PositionProps, SpaceProps } from 'styled-system';
export interface BoxProps extends SpaceProps, ColorProps, LayoutProps, FlexboxProps, PositionProps {
}
export declare const Box: import("styled-components").StyledComponent<"div", any, BoxProps, never>;
