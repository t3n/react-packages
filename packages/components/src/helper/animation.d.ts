import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
