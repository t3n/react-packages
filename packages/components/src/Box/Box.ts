import styled from 'styled-components';
import {
  color,
  ColorProps,
  flexbox,
  FlexboxProps,
  layout,
  LayoutProps,
  position,
  PositionProps,
  space,
  SpaceProps,
} from 'styled-system';

export interface BoxProps
  extends SpaceProps,
    ColorProps,
    LayoutProps,
    FlexboxProps,
    PositionProps {}

const Box = styled.div<BoxProps>`
  ${space}
  ${color}
  ${layout}
  ${flexbox}
  ${position}
`;

export default Box;
