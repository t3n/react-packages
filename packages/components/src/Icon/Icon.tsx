import React from 'react';
import styled from 'styled-components';
import {
  ColorProps,
  HeightProps,
  MarginProps,
  position,
  PositionProps,
  space,
  system,
  WidthProps,
} from 'styled-system';

export interface BaseIconProps
  extends React.SVGProps<SVGSVGElement>,
    MarginProps,
    PositionProps {
  component: React.FC<React.SVGProps<SVGSVGElement>>;
}

export interface IconProps
  extends Omit<BaseIconProps, 'width' | 'height' | 'fill'>,
    WidthProps,
    HeightProps {
  fill?: ColorProps['color'];
}

const BaseIcon = ({
  component: IconComponent,
  width,
  height,
  m,
  margin,
  mt,
  marginTop,
  mb,
  marginBottom,
  ml,
  marginLeft,
  mr,
  marginRight,
  my,
  mx,
  top,
  bottom,
  left,
  right,
  ...props
}: BaseIconProps) => (
  <IconComponent width={undefined} height={undefined} {...props} />
);

const Icon = styled(BaseIcon).attrs((props) => ({
  fill: 'text.primary',
  width: '1.5rem',
  ...props,
}))<IconProps>`
  ${system({
    fill: {
      property: 'fill',
      scale: 'colors',
    },
    width: true,
    height: true,
  })}
  position: relative;
  ${space}
  ${position}
`;

export default Icon;
