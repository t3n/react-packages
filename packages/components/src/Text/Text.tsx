import { ReactNode } from 'react';
import styled, { css } from 'styled-components';
import {
  color,
  space,
  width,
  textAlign,
  ColorProps,
  SpaceProps,
  WidthProps,
  TextAlignProps,
} from 'styled-system';
import { ThemeProps, composeTextStyle } from '@t3n/theme';

export interface TextProps extends ColorProps, SpaceProps, WidthProps {
  as?: 'p' | 'span';
  bold?: boolean;
  italic?: boolean;
  inline?: boolean;
  small?: boolean;
  secondary?: boolean;
  align?: TextAlignProps['textAlign'];
  children: ReactNode;
}

const font = ({ small, theme }: TextProps & ThemeProps) =>
  composeTextStyle({ textStyle: small ? 'small' : 'regular', theme });

const fontWeight = ({ bold }: TextProps) => (bold ? 'font-weight: bold;' : '');

const fontStyle = ({ italic }: TextProps) =>
  italic ? 'font-style: italic;' : '';

const textColor = ({
  color: colorProp,
  secondary,
  theme,
}: TextProps & ThemeProps) =>
  color({ color: secondary ? 'text.secondary' : colorProp, theme });

const align = ({ align: alignProp, theme }: TextProps & ThemeProps) =>
  textAlign({ textAlign: alignProp, theme });

export const textStyle = css<TextProps>`
  ${font}
  ${textColor}
  ${fontWeight}
  ${fontStyle}
  ${space}
  ${width}
  ${align}
`;

export const Text = styled.p.attrs(({ inline, as }: TextProps) => ({
  as: as || (inline ? 'span' : 'p'),
}))<TextProps>`
  ${textStyle}
`;
