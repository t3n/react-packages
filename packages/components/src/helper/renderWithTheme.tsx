/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
