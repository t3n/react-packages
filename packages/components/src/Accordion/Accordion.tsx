import React, { useState } from 'react';
import { MaterialArrowDropDown } from '@t3n/icons';
import styled from 'styled-components';
import { SpaceProps, space, MarginProps, color } from 'styled-system';
import { ThemeProps } from '@t3n/theme';
import { Icon } from '../Icon';
import { Heading } from '../Heading';
import { Box } from '../Box';

export interface AccordionProps extends MarginProps {
  title: string;
  initialOpen?: boolean;
}

const StyledAccordion = styled.div<SpaceProps>`
  border-radius: ${({ theme }: ThemeProps) => theme.border.radii[1]};

  ${({ theme }) => color({ theme, bg: 'shades.white' })};

  ${space};
`;

const StyledAccordionHeadBox = styled(Box)`
  display: flex;
  justify-content: space-between;
  cursor: pointer;

  :focus {
    outline: none;
  }
`;

const StyledIconBox = styled(Box)<{ collapsed: boolean }>`
  border-radius: 50%;
  height: 2rem;
  width: 2rem;

  ${Icon} {
    transform: ${({ collapsed }) =>
      collapsed ? 'rotate(180deg)' : 'rotate(0deg)'};
    transition: transform 0.2s ease-in;
  }
`;

export const Accordion: React.FC<AccordionProps> = ({
  children,
  title,
  initialOpen,
  ...rest
}) => {
  const [collapsed, setCollapsed] = useState(initialOpen || false);

  return (
    <StyledAccordion p={[3, 3, 4]} {...rest}>
      <StyledAccordionHeadBox
        onClick={() => setCollapsed(!collapsed)}
        onKeyDown={() => setCollapsed(!collapsed)}
        role="button"
        tabIndex={0}
      >
        <Heading as="h3" styleAs="h4" m={0}>
          {title}
        </Heading>

        <StyledIconBox bg="shades.grey244" collapsed={collapsed} ml={2}>
          <Icon
            width="2rem"
            height="2rem"
            fill="shades.grey143"
            component={MaterialArrowDropDown}
          />
        </StyledIconBox>
      </StyledAccordionHeadBox>
      {collapsed && (
        <Box mt={3} mb={5} pr={5}>
          {children}
        </Box>
      )}
    </StyledAccordion>
  );
};
