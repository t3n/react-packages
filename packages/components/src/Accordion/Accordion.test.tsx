import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent } from '@testing-library/dom';

import { MaterialAddAPhoto } from '@t3n/icons';

import Box from '../Box';
import { renderWithTheme } from '../helper/renderWithTheme';
import Icon from '../Icon';
import Text from '../Text';
import Accordion from './Accordion';

test('Accordion matches snapshot', () => {
  const { container } = renderWithTheme(
    <>
      <Accordion title="Some Accordion">
        <p>Content</p>
      </Accordion>
      <Accordion initialOpen title="Some Accordion">
        <p>Content</p>
      </Accordion>
    </>,
    {},
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('Accoridion can be initial open', () => {
  const { getByText } = renderWithTheme(
    <Accordion initialOpen title="Some Accordion">
      <p>Content</p>
    </Accordion>,
    {},
  );

  expect(getByText(/content/i)).toBeDefined();
});

test('Accoridion can be collapsed', () => {
  const { queryByText, getByRole } = renderWithTheme(
    <Accordion title="Some Accordion">
      <p>Content</p>
    </Accordion>,
    {},
  );

  expect(queryByText(/content/i)).toBeNull();
  fireEvent.click(getByRole('button'));
  expect(queryByText(/content/i)).not.toBeNull();
});

test('Accordion matches snapchot with JSX.Element', () => {
  const { container } = renderWithTheme(
    <Accordion
      title={
        <Box display="flex">
          <Icon component={MaterialAddAPhoto} mr={2} />
          <Text bold>Das hier ist ein JSX.Element</Text>
        </Box>
      }
    >
      <p>Content</p>
    </Accordion>,
    {},
  );

  expect(container.firstChild).toMatchSnapshot();
});
