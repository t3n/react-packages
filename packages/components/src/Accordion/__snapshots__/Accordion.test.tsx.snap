// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion machtes snapshot 1`] = `
.c5 {
  width: 2rem;
  height: 2rem;
  fill: #8F8F8F;
  position: relative;
}

.c2 {
  font-family: "adelle-sans",sans-serif;
  font-size: 1.125rem;
  line-height: 1.5em;
  margin: 0;
}

.c0 {
  border-radius: 4px;
  border: 1px solid #e8e8e8;
  background-color: #fff;
}

.c1 {
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

.c1:focus {
  outline: none;
}

.c3 {
  margin-left: 8px;
  background-color: #f4f4f4;
  border-radius: 50%;
  height: 2rem;
  width: 2rem;
}

.c3 .c4 {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: -webkit-transform 0.2s ease-in;
  -webkit-transition: transform 0.2s ease-in;
  transition: transform 0.2s ease-in;
}

.c6 .c4 {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
  -webkit-transition: -webkit-transform 0.2s ease-in;
  -webkit-transition: transform 0.2s ease-in;
  transition: transform 0.2s ease-in;
}

@media screen and (min-width:30rem) {
  .c2 {
    font-size: 1.125rem;
  }
}

@media screen and (min-width:48rem) {
  .c2 {
    font-size: 1.25rem;
  }
}

@media screen and (min-width:61.25rem) {
  .c2 {
    font-size: 1.25rem;
  }
}

@media screen and (min-width:90rem) {
  .c2 {
    font-size: 1.5rem;
  }
}

@media screen and (min-width:30rem) {
  .c2 {
    line-height: 1.5em;
  }
}

@media screen and (min-width:48rem) {
  .c2 {
    line-height: 1.333em;
  }
}

@media screen and (min-width:61.25rem) {
  .c2 {
    line-height: 1.333em;
  }
}

@media screen and (min-width:90rem) {
  .c2 {
    line-height: 1.333em;
  }
}

@media screen and (min-width:30rem) {

}

@media screen and (min-width:48rem) {

}

@media screen and (min-width:30rem) {
  .c1 {
    padding: 16px;
  }
}

@media screen and (min-width:48rem) {
  .c1 {
    padding: 24px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    role="button"
    tabindex="0"
  >
    <h3
      class="c2"
    >
      Some Accordion
    </h3>
    <div
      class="c3"
    >
      <svg
        class="c4 c5"
        fill="shades.grey143"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
